arch/x86/kvm/vmx/pkvm/hyp/iommu.c

```diff
@@ -322,36 +351,114 @@ static bool sync_shadow_context_entry(struct pgt_sync_data *sdata)
 {
 	struct context_entry *shadow_ce = sdata->shadow_ptep, tmp = {0};
 	struct context_entry *guest_ce = sdata->guest_ptep;
+	struct pkvm_iommu *iommu = pgt_to_pkvm_iommu(sdata->spgt);
+	struct pkvm_ptdev *ptdev;
+	struct pkvm_pgtable_cap cap;
 	bool updated = false;
-	u8 aw;
+	u8 tt, aw;
+	u16 bdf;
 
-	if (sdata->guest_ptep && sdata->shadow_pa) {
-		tmp.hi = guest_ce->hi;
-		tmp.lo = sdata->shadow_pa | (guest_ce->lo & 0xfff);
+	if (ecap_smts(sdata->iommu_ecap)) {
+		if (sdata->guest_ptep && sdata->shadow_pa) {
+			tmp.hi = guest_ce->hi;
+			tmp.lo = sdata->shadow_pa | (guest_ce->lo & 0xfff);
 
-		if (ecap_smts(sdata->iommu_ecap))
 			/* Clear DTE to make sure device TLB is disabled for security */
 			context_sm_clear_dte(&tmp);
-		else {
+		}
+	} else {
+		/*
+		 * In legacy mode, a context entry is a leaf entry responsible for
+		 * configuring the actual address translation for the given ptdev,
+		 * much like a PASID table entry in scalable mode. So the below logic
+		 * is quite similar to the logic in sync_shadow_pasid_table_entry()
+		 * for scalable mode.
+		 */
```

Comment:
Yep, it might be possible but definitely not easy (would need to introduce some new abstraction and significantly rework the code), it seems not worth it.

Reply:
BTW I added this comment mostly because `sync_shadow_pasid_table_entry()` includes some useful lengthy comments, like the explanation of iommu_audit_did() usage, and those explanations mostly apply here as well, so I wanted just to refer to `sync_shadow_pasid_table_entry()` altogether instead of duplicating those comments here.

Reply:
Okay